agent/ls.go

```diff
@@ -0,0 +1,136 @@
+package agent
+
+import (
+	"errors"
+	"net/http"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/httpapi"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+func (*agent) HandleLS(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	var query LSQuery
+	if !httpapi.Read(ctx, rw, r, &query) {
+		return
+	}
+
+	resp, err := listFiles(query)
+	if err != nil {
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			httpapi.Write(ctx, rw, http.StatusNotFound, codersdk.Response{
+				Message: "Directory does not exist",
+			})
+		case errors.Is(err, os.ErrPermission):
+			httpapi.Write(ctx, rw, http.StatusForbidden, codersdk.Response{
+				Message: "Permission denied",
+			})
+		default:
+			httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+				Message: err.Error(),
+			})
+		}
+		return
+	}
+
+	httpapi.Write(ctx, rw, http.StatusOK, resp)
+}
+
+func listFiles(query LSQuery) (LSResponse, error) {
+	var base string
+	switch query.Relativity {
+	case LSRelativityHome:
+		home, err := os.UserHomeDir()
+		if err != nil {
+			return LSResponse{}, xerrors.Errorf("failed to get user home directory: %w", err)
+		}
+		base = home
+	case LSRelativityRoot:
+		if runtime.GOOS == "windows" {
+			// TODO: Eventually, we could have a empty path with a root base
+			// return all drives.
+			// C drive should be good enough for now.
+			base = "C:\\"
+		} else {
+			base = "/"
+		}
+	default:
+		return LSResponse{}, xerrors.Errorf("unsupported relativity type %q", query.Relativity)
+	}
+
+	fullPath := append([]string{base}, query.Path...)
+	absolutePathString, err := filepath.Abs(filepath.Join(fullPath...))
+	if err != nil {
+		return LSResponse{}, xerrors.Errorf("failed to get absolute path: %w", err)
+	}
+
+	f, err := os.Open(absolutePathString)
+	if err != nil {
+		return LSResponse{}, xerrors.Errorf("failed to open directory: %w", err)
+	}
+	defer f.Close()
+
+	stat, err := f.Stat()
+	if err != nil {
+		return LSResponse{}, xerrors.Errorf("failed to stat directory: %w", err)
+	}
+
+	if !stat.IsDir() {
+		return LSResponse{}, xerrors.New("path is not a directory")
+	}
+
+	// `contents` may be partially populated even if the operation fails midway.
+	contents, _ := f.Readdir(-1)
```

Comment:
Go's documentation seems to recommend using `ReadDir` instead of `Readdir` which is pretty much the same API.

> Most clients are better served by the more efficient ReadDir method.